controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -944,118 +944,158 @@ private boolean isSubscriberToNonSubscriberTransition(final ReaderGroupConfigRec
-                            scope, stream, maxRecord, lowerBound, context, executor)
-                            .thenApply(gt -> {
-                            if (gt) {
-                                return maxRecord;
-                              } else {
-                                return lowerBound;
-                              }
-                            });
-                            })).orElse(CompletableFuture.completedFuture(null));
+                        } else { // greater than max. truncate at max. 
+                            // let there be data loss. its a lagging reader.. but if there is no streamcut in 
+                            // retentionset that satisfied the min and max criteria, then we should simply truncate at 
+                            // least at the lowerbound 
+                            return Optional.ofNullable(limits.getKey())
+                                           .filter(x -> currentSize - x.getRecordingSize() < retainedSizeLB)
+                                           .map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor)
+                                                                        .thenApply(StreamCutRecord::getStreamCut))
+                                    .orElse(CompletableFuture.completedFuture(lowerBound));
                         }
-                      }
+                    }
                 });
     }
 
     private CompletableFuture<Map<Long, Long>> getTruncationStreamCutByTimeLimit(String scope, String stream, OperationContext context,
                                                                                  RetentionPolicy policy, RetentionSet retentionSet,
                                                                                  Map<Long, Long> lowerBound) {
-        Map.Entry<StreamCutReferenceRecord, StreamCutReferenceRecord> limits =
-                getBoundStreamCuts(policy, retentionSet);
+        long currentTime = retentionClock.get().get();
+
+        Map.Entry<StreamCutReferenceRecord, StreamCutReferenceRecord> limits = getBoundStreamCuts(policy, retentionSet, 
+                x -> currentTime - x.getRecordingTime());
         // if subscriber lowerbound is ahead of streamcut corresponding to the max time and is behind stream cut for min time
         // from the retention set then we can safely truncate at lowerbound. Else we will truncate at the max time bound if it
         // exists
-        // 1. if LB > min => truncate at min
-        // 2. if LB < max => truncate at max
-        // 3. if LB < min && LB > max => truncate at LB
-        // 4. if LB < min && overlaps max => truncate at max
+        // 1. if LB is ahead of min => truncate at min
+        // 2. if LB is behind max => truncate at max
+        // 3. if LB is behind min && LB is ahead of max => truncate at LB
+        // 4. if LB is behind min && overlaps max => truncate at max
```

Comment:
`4. if LB is behind min && overlaps max => truncate at max`
Here, why not truncate at `LB` instead of truncate at max?

Reply:
if it overlaps with max, it means it has some data which is older than max while some data which is newer than max.
since our limit on time is satisfied by `max` we wouldnt want to retain older data than that.. hence truncating at max which satisfies the criteria more accurately than truncating at LB